contracts/mainnet/connectors/morpho-blue/main.sol

```diff
@@ -0,0 +1,918 @@
+		_eventParam = abi.encode(
+			_marketParams,
+			_assets,
+			_onBehalf,
+			_getId,
+			_setId
+		);
+	}
+
+	/**
+	 * @notice Handles the collateral withdrawals.
+	 * @dev The market to withdraw assets from. (For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param _marketParams The market to withdraw assets from.
+	 * @param _assets The amount of assets to withdraw. (For max: `uint256(-1)`)
+	 * @param _getId ID to retrieve amt.
+	 * @param _setId ID stores the amount of tokens deposited.
+	 */
+	function withdrawCollateral(
+		MarketParams memory _marketParams,
+		uint256 _assets,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _assets);
+
+		_marketParams = updateTokenAddresses(_marketParams);
+
+		// If amount is max, fetch collateral value from Morpho's contract
+		if (_amt == type(uint256).max) {
+			bytes32 _id = id(_marketParams);
+			Position memory _pos = MORPHO_BLUE.position(_id, address(this));
+			_amt = _pos.collateral;
+		}
+
+		MORPHO_BLUE.withdrawCollateral(
+			_marketParams,
+			_amt,
+			address(this),
+			address(this)
+		);
+
+		convertWethToEth(
+			_marketParams.collateralToken == ethAddr,
+			TokenInterface(wethAddr),
+			_amt
+		);
+
+		setUint(_setId, _amt);
+
+		_eventName = "LogWithdrawCollateral((address,address,address,address,unit256),uint256,uint256,uint256)";
+		_eventParam = abi.encode(_marketParams, _amt, _getId, _setId);
+	}
+
+	/**
+	 * @notice Handles the withdrawal of collateral by a user from a specific market of a specific amount.
+	 * @dev The market to withdraw assets from. (For ETH: 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)
+	 * @param _marketParams The market to withdraw assets from.
+	 * @param _assets The amount of assets to withdraw. (For max: `uint256(-1)`)
+	 * @param _onBehalf The address that already deposited position.
+	 * @param _getId ID to retrieve amt.
+	 * @param _setId ID stores the amount of tokens deposited.
+	 */
+	function withdrawCollateralOnBehalf(
+		MarketParams memory _marketParams,
+		uint256 _assets,
+		address _onBehalf,
+		address _receiver,
+		uint256 _getId,
+		uint256 _setId
+	)
+		external
+		payable
+		returns (string memory _eventName, bytes memory _eventParam)
+	{
+		uint256 _amt = getUint(_getId, _assets);
+
+		_marketParams = updateTokenAddresses(_marketParams);
+
+		// If amount is max, fetch collateral value from Morpho's contract
+		if (_amt == type(uint256).max) {
+			bytes32 _id = id(_marketParams);
+			Position memory _pos = MORPHO_BLUE.position(_id, _onBehalf);
+			_amt = _pos.collateral;
+		}
+
+		MORPHO_BLUE.withdrawCollateral(
+			_marketParams,
+			_amt,
+			_onBehalf,
+			_receiver
+		);
+
+		if (_receiver == address(this))
+			convertWethToEth(
+				_marketParams.collateralToken == ethAddr,
+				TokenInterface(wethAddr),
+				_amt
+			);
```

Comment:
Conversion is only happening in case the token addresses sent from our UI for loan (or collateral) is ETH (0xeee...). We allow users to use ETH tokens for the WETH markets and do the Weth to Eth conversion internally.